auction-server/src/opportunity/service/add_opportunity.rs

```diff
@@ -24,23 +25,49 @@ pub struct AddOpportunityInput<T: entities::OpportunityCreate> {
     pub opportunity: T,
 }
 
+type OpportunityType<T> = <<T as ChainType>::InMemoryStore as InMemoryStore>::Opportunity;
+type OpportunityCreateType<T> = <OpportunityType<T> as entities::Opportunity>::OpportunityCreate;
+
+#[derive(Debug, Clone)]
+enum OpportunityAction<T: entities::Opportunity> {
+    Add,
+    Refresh(T),
+    Ignore,
+}
+
 impl<T: ChainType> Service<T>
 where
     Service<T>: Verification<T>,
 {
+    async fn assess_action(
+        &self,
+        opportunity: &OpportunityCreateType<T>,
+    ) -> OpportunityAction<OpportunityType<T>> {
+        let opportunities = self
+            .repo
+            .get_in_memory_opportunities_by_key(&opportunity.get_key())
+            .await;
+        for opp in opportunities.iter() {
+            let comparison = opp.compare(opportunity);
+            if let entities::OpportunityComparison::Duplicate = comparison {
+                return OpportunityAction::Ignore;
+            }
+            if let entities::OpportunityComparison::NeedsRefresh = comparison {
+                return OpportunityAction::Refresh(opp.clone());
```

Comment:
clone inspector nit: we can remove the clone and use `into_iter` on the for loop :D